man/man3/wdLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 21, 2005
.Dt wdLib 3
.Os pocoLibs
.Sh NAME 
.Nm wdLib
.Nd watchdog library for pocoLibs
.Sh SYNOPSIS
.Fd "#include <portLib.h>"
.Fd "#include <wdLib.h>"
.Ft STATUS
.Fn wdLibInit void
.Ft WDOG_ID
.Fn wdCreate void
.Ft STATUS
.Fn wdDelete "WDOG_ID wdId"
.Ft STATUS
.Fn wdStart "WDOG_ID wdID" "int delay" "FUNCPTR routine" \
"long parameter"
.Ft STATUS
.Fn wdCancel "WDOG_ID wdID"
.Sh DESCRIPTION
.Fn wdLibInit
initializes the watch dog facility of 
.Nm pocoLibs
.Fn wdCreate
creates a new inactive watch dog. 
.Pp
.Fn wdDelete
destroys an exising watch dog.
.Pp
.Fn wdStart
starts the timer associated with a watch dog. 
.Fa wdId
is a watch dog identifier as returned by
.Fn wdCreate . 
.Fa delay 
is the number of ticks until the watch dog expires. 
.Fa routine
is a pointer to a function that will be called when the timer
expires. 
.Fa parameter 
is an integer value that will be passed to the 
.Fa routine
when it will be called.
.Pp
.Fn wdCancel
cancels the active watchDog designated by
.Fa wdId .
.Sh RETURN VALUES
.Fn wdCreate
returns the identifier of the newly created watchdor or 
.Dv NULL
if an error occured. 
.Pp
.Fn wdLibInit ,
.Fn wdDelete ,
.Fn wdStart 
and 
.Fn wdCancel
return
.Dv OK
or
.Dv ERROR
if an error ocurred. All functions set the task error code in case of
errors. 
.Sh ERRORS
The error code of the current task can be set to:
.Bl -tag -width Er
.It Bq Er S_memLib_NOT_ENOUGH_MEMORY
The allocation of memory for a new watchog failed.
.It Bq Eq S_objLib_OBJ_ID_ERROR
The 
.Fa wdId
passed to a function is not valid.
.Sh SEE ALSO
.Xr sysLib 3 ,
.Xr timerLib 3 .


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/portLib/os/posix/wdLib.c
/*
 * Copyright (c) 1999-2005 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "pocolibs-config.h"

#include "portLib.h"

#include <signal.h>
#include <stdlib.h>
#include <pthread.h>

#include "errnoLib.h"
#include "wdLib.h"
static const H2_ERROR wdLibH2errMsgs[] = WD_LIB_H2_ERR_MSGS;

WDOG_ID wdList = NULL;

static pthread_mutex_t wdMutex = PTHREAD_MUTEX_INITIALIZER;

STATUS
wdLibInit(void)
{
    /* record error msgs */
    h2recordErrMsgs("wdLibInit", "wdLib", M_wdLib, 			
		    sizeof(wdLibH2errMsgs)/sizeof(H2_ERROR), 
		    wdLibH2errMsgs);

    return OK;
}

WDOG_ID
wdCreate(void)
{
    WDOG_ID wd;
    sigset_t set;

    wd = (WDOG_ID)malloc(sizeof(struct wdog));
    if (wd == NULL) {
	errnoSet(S_wdLib_NOT_ENOUGH_MEMORY);
	return NULL;
    }
    wd->delay = -1;
    wd->pRoutine = NULL;
    wd->magic = WDLIB_MAGIC;

    /* Block clock signal */
    sigemptyset(&set);
    sigaddset(&set, SIGALRM);
    sigprocmask(SIG_BLOCK, &set, NULL);

    pthread_mutex_lock(&wdMutex);

    /* Insert into wdList */
    wd->next = wdList;
    wdList = wd;
    pthread_mutex_unlock(&wdMutex);
    sigprocmask(SIG_UNBLOCK, &set, NULL);
    return wd;
}

STATUS
wdDelete(WDOG_ID wdId)
{
    WDOG_ID wd;
    sigset_t set;

    if (wdId == NULL || wdId->magic != WDLIB_MAGIC) {
	errnoSet(S_wdLib_ID_ERROR);
	return ERROR;
    }
    /* Block clock signal */
    sigemptyset(&set);
    sigaddset(&set, SIGALRM);
    sigprocmask(SIG_BLOCK, &set, NULL);

    /* Remove from wdList */
    pthread_mutex_lock(&wdMutex);
    if (wdId == wdList) {
	wdList = wdId->next;
    } else {
	for (wd = wdList; wd->next != NULL; wd = wd->next) {
	    if (wd->next == wdId) {
		wd->next = wdId->next;
		break;
	    }
	} /* for */
    }
    free(wdId);
    pthread_mutex_unlock(&wdMutex);
    sigprocmask(SIG_UNBLOCK, &set, NULL); /* unblock clock sig */
    return OK;
}

STATUS
wdStart(WDOG_ID wdId, int delay, FUNCPTR pRoutine, long parameter)
{
    sigset_t set;

    if (wdId == NULL || wdId->magic != WDLIB_MAGIC) {
	errnoSet(S_wdLib_ID_ERROR);
	return ERROR;
    }
    /* Block clock signal */
    sigemptyset(&set);
    sigaddset(&set, SIGALRM);
    sigprocmask(SIG_BLOCK, &set, NULL);

    /* XXX */
    /* mutex_lock n'est pas aync signal safe, or wdStart() doit l'etre */
    /* Il faut trouver une autre solution */
    pthread_mutex_lock(&wdMutex);
    wdId->delay = delay;
    wdId->pRoutine = pRoutine;
    wdId->parameter = parameter;
    pthread_mutex_unlock(&wdMutex);

    sigprocmask(SIG_UNBLOCK, &set, NULL);
    return OK;
}

STATUS
wdCancel(WDOG_ID wdId)
{
    sigset_t set;

    if (wdId == NULL || wdId->magic != WDLIB_MAGIC) {
	errnoSet(S_wdLib_ID_ERROR);
	return ERROR;
    }
    /* Block clock signal */
    sigemptyset(&set);
    sigaddset(&set, SIGALRM);
    sigprocmask(SIG_BLOCK, &set, NULL);

    pthread_mutex_lock(&wdMutex);
    wdId->delay = 0;
    wdId->pRoutine = NULL;
    pthread_mutex_unlock(&wdMutex);

    sigprocmask(SIG_UNBLOCK, &set, NULL); /* unblock clock sig */
    return OK;
}



	
